src/events/guildAuditLogEntryCreate.ts

```diff
@@ -0,0 +1,224 @@
+import { AuditLogEvent, EmbedBuilder } from 'discord.js';
+import { sendEventLogMessage } from '@/util';
+import type { Guild, User, GuildAuditLogsEntry } from 'discord.js';
+
+export default async function guildAuditLogEntryCreateHandler(auditLogEntry: GuildAuditLogsEntry, guild: Guild): Promise<void> {
+	switch (auditLogEntry.action) { 
+		case AuditLogEvent.MemberUpdate: {
+
+			// * Unsure why just checking the action type doesn't narrow the type here, probably too much type inference going on?
+			const user = (auditLogEntry as unknown as GuildAuditLogsEntry<AuditLogEvent.MemberUpdate>).target;
```

Comment:
So I tried this, unfortunately the types are _really messed up_.

In theory, by asserting that `auditLogEntry.action === ActionLogEvent.MemberUpdate` (which we're doing in the switch) then the type of `auditLogEntry` should be inferred as `GuildAuditLogsEntry<ActionLogEvent.MemberUpdate>`. Unfortunately the type of '`auditLogEntry` doesn't get refined by the `switch` it doesn't even get refined if I change this into a chained `if`.

I tried creating a custom type predicate and that wasn't working, but I've just tried again and it works, so fuck me I guess. I'll update

Reply:
Can you post the actual TypeScript errors your are getting here? I'm trying to replicate the issue but I can't my tests are narrow the type down correctly:

![Screenshot from 2024-07-19 19-12-34](https://github.com/user-attachments/assets/390f9e33-dd99-40c9-ab0a-a195929d242b)

Reply:
In your example here, the type of `user` isn't getting narrowed? It's still any of that set of union types. I want it to be a `User | null` which is what it would be in a `GuildAuditLogsEntry<AuditLogEvent.MemberUpdate>`

I've changed from the switch to a series of `if` / `else`s using the type predicate to narrow the type so that the type of the event is narrowed, giving the right types for the target

![image](https://github.com/user-attachments/assets/150d3f63-b615-42c6-986b-2136fbc9dcb1)

Example of the error:
![image](https://github.com/user-attachments/assets/7c14f4ca-f4e9-4bbf-9d99-394671d14399)

Reply:
I was referring to the `auditLogEntry` type, not the `user` type. The type assertions you're doing are on `auditLogEntry` in order to get the `target`, but it already has access to the `target` field at this point

If what you actually want is to narrow the `user` type, you can simply do that on the `user` itself:

```ts
import {
	AuditLogEvent,
	type GuildAuditLogsEntry,
	type User
} from 'discord.js';

function test(auditLogEntry: GuildAuditLogsEntry) {
	switch (auditLogEntry.action) {
		case AuditLogEvent.MemberUpdate:
			const user = auditLogEntry.target as User | null;

			if (!user) {
				return;
			}

			handleUser(user);
			break;
	}
}

function handleUser(user: User) {
	// Do something
}
```

The concern I was pointing out was mostly centered around the use of `as unknown as GuildAuditLogsEntry<AuditLogEvent.MemberUpdate>`, which is hard to parse at a glance and (not your fault, it's the libs fault), honestly just ugly. Asserting on `user` is much more clear and shorter

Reply:
Are you ok with the way that I've done this currently? Using a type predicate to narrow the audit log type, then I don't need to cast `user` at all

e.g.
```ts
	if (logIsForEvent(auditLogEntry, AuditLogEvent.MemberUpdate)) {
		const user = auditLogEntry.target;
		if (!user) {
			return;
		}
                ...
```

Reply:
Yes that's fine :+1: you say "currently", was this already in place? I think that the GitHub mobile app may have been playing tricks on me last night when I did this review...

Reply:
Yeah, it was already like this 😂